### PR TITLE
Fix Index typehints and move class

### PR DESCRIPTION
**Pull request by sfc-gh-samdesai1**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1481037

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Fix some of the typehints in the index class to correctly identify return and input types

### CODE REVIEW

**Review by sfc-gh-vbudati — APPROVED** [review]
Thanks Samedh!

**Review by sfc-gh-yzou — APPROVED** [review]
@sfc-gh-samdesai1 please try to generate the api doc and see if the type are set correctly

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 1]
```diff
@@ -42,8 +46,7 @@
 class Index:
     def __init__(
         self,
-        # TODO: SNOW-1481037 : Fix typehints for index constructor, set_query_compiler and set_local_index
-        data: ArrayLike | Any = None,
+        data: ArrayLike | SnowflakeQueryCompiler = None,
```
Comment (sfc-gh-joshi) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:49]:
```suggestion
        data: ArrayLike | SnowflakeQueryCompiler | None = None,
```
You should add the extra `None` annotation to make clear this is optional.

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 2]
```diff
@@ -109,8 +112,7 @@ def __init__(
 
     def set_query_compiler(
         self,
-        # TODO: SNOW-1481037 : Fix typehints for index constructor, set_query_compiler and set_local_index
-        data: ArrayLike | Any = None,
+        data: ArrayLike | SnowflakeQueryCompiler = None,
```
Comment (sfc-gh-joshi) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:115]:
```suggestion
        data: ArrayLike | SnowflakeQueryCompiler | None = None,
```

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 3]
```diff
@@ -140,8 +137,7 @@ def set_query_compiler(
 
     def set_local_index(
         self,
-        # TODO: SNOW-1481037 : Fix typehints for index constructor, set_query_compiler and set_local_index
-        data: ArrayLike | Any = None,
+        data: ArrayLike | SnowflakeQueryCompiler = None,
```
Comment (sfc-gh-joshi) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:140]:
```suggestion
        data: ArrayLike | SnowflakeQueryCompiler | None = None,
```

**Review by sfc-gh-joshi — CHANGES_REQUESTED** [review]
A couple of nits regarding optional types. A couple more things:

Can you fix the type annotations on these three methods?
- `is_lazy_check`: just making it go `Callable -> Callable` is fine, as long as it's clear that this is an annotation. This also should be moved out of the class to the top level of the file, since it's neither an instance nor class method (first argument isn't `self`/`cls`).
- `duplicated` (currently returns Any)
- `str` (return type should be `native_pd.core.strings.accessor.StringMethods` rather than `str`)

I also noticed there's a function that uses `Self` in the return type; we should explicitly use `Index` instead. If you want to use `Self`, you should replace all the other type hint references to `Index` as well, and change the import to come from the native `typing` module rather than `pandas._typing`.

**Review by sfc-gh-joshi — APPROVED** [review]

**Pull request merged**